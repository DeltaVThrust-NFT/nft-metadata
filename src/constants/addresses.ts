export const RARIBLE_TOKEN_ADDRESS =
  '0x60F80121C31A0d46B5279700f9DF786054aa5eE5'

export const BORED_APE_TOKEN_ADDRESS =
  '0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D'

export const VEE_FRIENDS_TOKEN_ADDRESS =
  '0xa3AEe8BcE55BEeA1951EF834b99f3Ac60d1ABeeB'

export const SORARE_TOKEN_ADDRESS = '0x629a673a8242c2ac4b7b8c5d8735fbeac21a6205'

export const SUPER_YETI_TOKEN_ADDRESS =
  '0x3F0785095A660fEe131eEbcD5aa243e529C21786'

export const CRYPTOVOXELS_TOKEN_ADDRESS =
  '0x79986aF15539de2db9A5086382daEdA917A9CF0C'

export const BONSAI_TOKEN_ADDRESS = '0xeC9C519D49856Fd2f8133A0741B4dbE002cE211b'

export const SUPERRARE_TOKEN_ADDRESS =
  '0xb932a70A57673d89f4acfFBE830E8ed7f75Fb9e0'

export const HASHMASKS_TOKEN_ADDRESS =
  '0xC2C747E0F7004F9E8817Db2ca4997657a7746928'

export const DECENTRALAND_TOKEN_ADDRESS =
  '0xF87E31492Faf9A91B02Ee0dEAAd50d51d56D5d4d'

export const ZORA_TOKEN_ADDRESS = '0xabEFBc9fD2F806065b4f3C237d4b59D9A97Bcac7'
export const ZORA_RINKEBY_TOKEN_ADDRESS =
  '0x7C2668BD0D3c050703CEcC956C11Bd520c26f7d4'

export const MAKERSPLACE_TOKEN_ADDRESS =
  '0x2A46f2fFD99e19a89476E2f62270e0a35bBf0756'

export const ART_BLOCKS_CURATED_TOKEN_ADDRESS =
  '0x059EDD72Cd353dF5106D2B9cC5ab83a52287aC3a'

export const ART_BLOCKS_TOKEN_ADDRESS =
  '0xa7d8d9ef8D8Ce8992Df33D8b8CF4Aebabd5bD270'

export const FOUNDATION_TOKEN_ADDRESS =
  '0x3B3ee1931Dc30C1957379FAc9aba94D1C48a5405'

export const WRAPPED_CRYPTOPUNKS_TOKEN_ADDRESS =
  '0xb7F7F6C52F2e2fdb1963Eab30438024864c313F6'

export const BLITMAP_TOKEN_ADDRESS =
  '0x8d04a8c79cEB0889Bdd12acdF3Fa9D207eD3Ff63'

export const KNOWN_ORIGIN_TOKEN_ADDRESS =
  '0xFBeef911Dc5821886e1dda71586d90eD28174B7d'

export const HEAVEN_COMPUTER_TOKEN_ADDRESS =
  '0x80ADB36595239fe918c7D118C1F81e07d070801a'

export const MEEBITS_TOKEN_ADDRESS =
  '0x7Bd29408f11D2bFC23c34f18275bBf23bB716Bc7'

export const HOLLY_PLUS_TOKEN_ADDRESS =
  '0x6688Ee4E6e17a9cF88A13Da833b011E64C2B4203'

export const LOOT_TOKEN_ADDRESS = '0xFF9C1b15B16263C61d017ee9F65C50e4AE0113D7'

export const SETTLEMENTS_TOKEN_ADDRESS =
  '0xdEcC60000ba66700a009b8F9F7D82676B5cfA88A'

export const BLITNAUT_TOKEN_ADDRESS =
  '0x448f3219CF2A23b0527A7a0158e7264B87f635Db'

export const SOLVENCY_TOKEN_ADDRESS =
  '0x82262bFba3E25816b4C720F1070A71C7c16A8fc4'

export const BLOCKPARY_TOKEN_ADDRESS =
  '0x2Fb704d243cFA179fFaD4D87AcB1D36bcf243a44'

export const MANNYS_GAME_TOKEN_ADDRESS =
  '0x2bd58A19C7E4AbF17638c5eE6fA96EE5EB53aed9'

export const ALGOLITE_TOKEN_ADDRESS =
  '0x5c685a3EBC751F36B2123C25F5c464D3b9964AfC'

export const HUMANOID_TOKEN_ADDRESS =
  '0x3a5051566b2241285BE871f650C445A88A970edd'
