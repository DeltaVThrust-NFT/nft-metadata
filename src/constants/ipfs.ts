export const IPFS_IO_GATEWAY = 'https://gateway.ipfs.io'
