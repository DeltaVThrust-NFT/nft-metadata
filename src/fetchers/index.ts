import { fetchUnderlyingContractData } from './defaultFetcher'
import { getAddress } from '@ethersproject/address'

export function fetcherLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    default:
      return fetchUnderlyingContractData
  }
}
