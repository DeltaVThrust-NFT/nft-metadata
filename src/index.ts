export * from './constants'
export * from './metadata'
export * from './agent'
export * from './uri'
