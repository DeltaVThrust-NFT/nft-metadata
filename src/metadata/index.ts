export * from './fetcher'
export * from './hashmasks'
export * from './loot'
export * from './normaliser'
export * from './zora'
