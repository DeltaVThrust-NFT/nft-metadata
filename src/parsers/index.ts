import { getAddress } from '@ethersproject/address'
import { parseGenericMetadata } from './openseaMetadataParser'
import {
  ART_BLOCKS_CURATED_TOKEN_ADDRESS,
  ART_BLOCKS_TOKEN_ADDRESS,
  HASHMASKS_TOKEN_ADDRESS,
  MAKERSPLACE_TOKEN_ADDRESS,
  ZORA_TOKEN_ADDRESS,
} from '../constants'
import { parseZoraMetadata } from './zoraMetadataParser'
import { parseMakersplaceMetadata } from './makersplaceMetadataParser'
import { parseHashmasksMetadata } from './hashmasksMetadataParser'
import { parseArtblocksMetadata } from './artblocksMetadataParser'

export function parserLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    case ART_BLOCKS_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case ART_BLOCKS_CURATED_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case HASHMASKS_TOKEN_ADDRESS:
      return parseHashmasksMetadata
    case MAKERSPLACE_TOKEN_ADDRESS:
      return parseMakersplaceMetadata
    case ZORA_TOKEN_ADDRESS:
      return parseZoraMetadata
    default:
      return parseGenericMetadata
  }
}
