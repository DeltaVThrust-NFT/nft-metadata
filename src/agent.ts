import { JsonRpcProvider } from '@ethersproject/providers'
import { fetchOwnerOf, getSupportedInterfaces } from './utils/contracts'
import { Erc721Factory } from '@zoralabs/core/dist/typechain'
import { Fetcher, fetcherLookup, FetcherResponse } from './fetchers'
import { fetchMetadata, fetchMimeType } from './utils/fetch'
import { getIPFSUrl } from './utils/ipfs'
import { Parser, parserLookup } from './parsers'

export interface TokenPKDto {
  networkId: number
  tokenAddress: string
  tokenId: string
}

export interface NftMetadata extends TokenPKDto {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string
  creatorAddress?: string

  tokenURI: string
  tokenURL: string
  tokenURLMimeType: string

  contentURL?: string
  contentURLMimeType?: string

  imageURL?: string
  imageURLMimeType?: string

  externalLink?: string

  attributes?: Record<string, any>[]
}

export enum Network {
  MAINNET = 1,
  RINKEBY = 4,
}

export type AgentNetworkConfiguration = {
  [key: number]: JsonRpcProvider
}

export type AgentConfiguration = {
  ipfsGateway?: string
  fetchTimeout?: number
  providers: AgentNetworkConfiguration
  parsers?: {
    [key: string]: Parser
  }
  fetchers?: {
    [key: string]: Fetcher
  }
}

export type TokenContractData = FetcherResponse & TokenPKDto & Record<any, any>
export type TokenData = { metadata: Record<any, any>; contentType: string }

export class Agent {
  constructor(private readonly options: AgentConfiguration) {}

  private getProviderForNetwork(number: Network): JsonRpcProvider {
    const provider = this.options.providers?.[number]
    if (!provider) {
      throw new Error('No provider for network')
    }
    return provider
  }

  public async supportedInterfaces(networkId: Network, tokenAddress: string) {
    const provider = this.getProviderForNetwork(networkId)
    return getSupportedInterfaces(provider, tokenAddress)
  }

  public async fetchContractData(
    networkId: Network,
    tokenAddress: string,
    tokenId: string,
  ) {
    const provider = this.getProviderForNetwork(networkId)
    const { erc721, erc721Metadata, erc721Enumerable } =
      await this.supportedInterfaces(networkId, tokenAddress)

    if (!erc721) {
      throw new Error(
        `${tokenAddress} on network: ${networkId} does not support ERC721 standard`,
      )
    }

    const erc721Contract = Erc721Factory.connect(tokenAddress, provider)
    const ownerAddress = await fetchOwnerOf(
      erc721Contract,
      tokenId,
      // Can catch if contract can make a total supply call
      // to handle open zeppelin burn
      erc721Enumerable,
    )

    let tokenURI: string | undefined
    if (erc721Metadata) {
      tokenURI = await erc721Contract.tokenURI(tokenId)
    }

    const baseData = {
      networkId,
      tokenAddress,
      tokenId,
      tokenURI,
      ownerAddress,
    }

    // bespoke fetches for other underlying data
    const fetcher = fetcherLookup(tokenAddress, this.options.fetchers)
    if (fetcher) {
      const data = await fetcher({
        provider,
        tokenAddress,
        tokenId,
      })

      return {
        ...baseData,
        ...data,
      }
    }

    return baseData
  }

  public async parseMetadata(
    contractData: TokenContractData,
    tokenData: TokenData,
  ) {
    const metadata = tokenData.metadata

    let meta: Partial<NftMetadata> = {}

    if (contractData.tokenURI) {
      meta.tokenURL = getIPFSUrl(
        contractData.tokenURI,
        this.options.ipfsGateway,
      )
      meta.tokenURLMimeType = tokenData.contentType
    }

    if (metadata.image) {
      meta.imageURL = getIPFSUrl(metadata.image, this.options.ipfsGateway)
      meta.contentURL = meta.imageURL
    }

    if (metadata.animation_url) {
      meta.contentURL = getIPFSUrl(
        metadata.animation_url,
        this.options.ipfsGateway,
      )
    }

    const {
      name,
      description,
      attributes,
      traits,
      external_url: externalURL,
      mimeType,
    } = metadata

    if (meta.contentURL) {
      meta.contentURLMimeType = await fetchMimeType(
        meta.contentURL,
        {
          timeout: this.options.fetchTimeout,
        },
        mimeType,
      )
    }

    if (meta.imageURL && meta.imageURL !== meta.contentURL) {
      meta.imageURLMimeType = await fetchMimeType(meta.imageURL, {
        timeout: this.options.fetchTimeout,
      })
    } else {
      delete meta.imageURL
    }

    const baseMeta = {
      metadata,
      ...meta,
      ...(name && { name }),
      ...(description && { description }),
      ...(attributes && { attributes }),
      ...(traits && { attributes: traits }),
      ...(externalURL && { externalURL }),
    }

    // Return any extra fields unparseable by defaults
    // @TODO - @ethandaya move this behaviour to a dynamic registry
    const parser = parserLookup(contractData.tokenAddress, this.options.parsers)
    if (parser) {
      const meta = await parser({
        provider: this.getProviderForNetwork(contractData.networkId),
        contractData,
        tokenData,
        ipfsGateway: this.options.ipfsGateway,
        fetchTimeout: this.options.fetchTimeout,
        baseMeta,
      })
      return {
        ...baseMeta,
        ...meta,
      }
    }

    return baseMeta
  }

  public async fetchAndParseTokenData(
    networkId: number,
    tokenAddress: string,
    tokenId: string,
  ) {
    const contractData = await this.fetchContractData(
      networkId,
      tokenAddress,
      tokenId,
    )

    if (!contractData.tokenURI) {
      return { ...contractData }
    }

    const tokenData = await fetchMetadata(contractData.tokenURI, {
      timeout: this.options.fetchTimeout,
      ipfsGateway: this.options.ipfsGateway,
    })

    const metadata = await this.parseMetadata(contractData, tokenData)

    return {
      ...contractData,
      ...metadata,
    }
  }
}
