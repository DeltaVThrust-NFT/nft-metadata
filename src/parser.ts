import { JsonRpcProvider } from '@ethersproject/providers'
import { getAddress } from '@ethersproject/address'
import { fetcherLookup } from './fetchers'
import { parserLookup } from './parsers'

export interface NftMetadata {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string

  tokenURI: string
  contentURI: string

  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Parser {
  constructor(private readonly provider: JsonRpcProvider) {}

  public async fetchUnderlyingContractData(
    contractAddress: string,
    tokenId: string,
  ) {
    const fetcher = fetcherLookup(contractAddress)
    return fetcher(this.provider, contractAddress, tokenId)
  }

  public async parseTokenMetadata(
    contractAddress: string,
    tokenId: string,
    tokenURI: string,
  ) {
    const parser = parserLookup(contractAddress)
    return parser(this.provider, contractAddress, tokenId, tokenURI)
  }

  public async fetchAndParseTokenMeta(
    tokenAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const safeAddress = getAddress(tokenAddress)

    const { tokenURI, ownerAddress } = await this.fetchUnderlyingContractData(
      safeAddress,
      tokenId,
    )

    const parsedMetadata = await this.parseTokenMetadata(
      safeAddress,
      tokenId,
      tokenURI,
    )

    return {
      tokenURI,
      ownerAddress,
      ...parsedMetadata,
    }
  }
}
