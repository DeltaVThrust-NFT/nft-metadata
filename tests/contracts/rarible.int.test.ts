import { Parser } from '../../src/parser'
import { RARIBLE_TOKEN_ADDRESS } from '../../src/constants/addresses'
import RARIBLE_METADATA_STUB from '../mock-reponses/contracts/rarible/301_962.json'
import { testProvider } from '../mock-reponses/setupProvider'
import { isAddress } from '@ethersproject/address'

const RARIBLE_MAINNET_MOCK = {
  input: {
    tokenId:
      '30105810670306411381802272225292624615592980350863712708305530245480406056962',
    tokenAddress: RARIBLE_TOKEN_ADDRESS,
  },
  output: {
    metadata: RARIBLE_METADATA_STUB,
    name: RARIBLE_METADATA_STUB.name,
    description: RARIBLE_METADATA_STUB.description,
    tokenURI:
      'https://ipfs.io/ipfs/QmfHrsEpXXrvi2dTTNake723kkMapDQXeuYDsVmSRdsQNH',
    contentURI: RARIBLE_METADATA_STUB.image,
    externalURL: RARIBLE_METADATA_STUB.external_url,
    attributes: RARIBLE_METADATA_STUB.attributes,
  },
}

describe('Rarible ERC721', () => {
  const parser = new Parser(testProvider)

  beforeEach(() => {
    jest.setTimeout(10000)
  })

  it('should be able to fetch and parse metadata for a rarible erc721', async () => {
    const { ownerAddress, ...meta } = await parser.fetchAndParseTokenMeta(
      RARIBLE_MAINNET_MOCK.input.tokenAddress,
      RARIBLE_MAINNET_MOCK.input.tokenId,
    )
    expect(meta).toStrictEqual(RARIBLE_MAINNET_MOCK.output)
    expect(isAddress(ownerAddress)).toBeTruthy()
  })
})
