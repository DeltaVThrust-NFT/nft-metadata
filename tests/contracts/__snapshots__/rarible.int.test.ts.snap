// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rarible ERC721 should be able to fetch and parse metadata for token id: 30105810670306411381802272225292624615592980350863712708305530245480406056962 on network: 1 1`] = `
Object {
  "attributes": Array [
    Object {
      "trait_type": "Creator's Handle",
      "value": "mynamebrody",
    },
    Object {
      "trait_type": "Permalink",
      "value": "https://www.instagram.com/p/CLZ-xK7liM4/",
    },
    Object {
      "trait_type": "Created Date",
      "value": "2021-02-17T20:00:18+0000",
    },
    Object {
      "trait_type": "Platform",
      "value": "Instagram",
    },
  ],
  "contentURL": "https://ipfs.fleek.co/ipfs/QmcPsUASuDWuxbpoXPjgBrS6UHuxEdp3ZFwM88dKYsvTVH",
  "contentURLMimeType": "image/jpeg",
  "description": "❄️ Winter Daze ☀️ #aryaeverywhere",
  "externalURL": "https://coconft.com/p/17877779234167542",
  "metadata": Object {
    "attributes": Array [
      Object {
        "trait_type": "Creator's Handle",
        "value": "mynamebrody",
      },
      Object {
        "trait_type": "Permalink",
        "value": "https://www.instagram.com/p/CLZ-xK7liM4/",
      },
      Object {
        "trait_type": "Created Date",
        "value": "2021-02-17T20:00:18+0000",
      },
      Object {
        "trait_type": "Platform",
        "value": "Instagram",
      },
    ],
    "description": "❄️ Winter Daze ☀️ #aryaeverywhere",
    "external_url": "https://coconft.com/p/17877779234167542",
    "image": "https://gateway.pinata.cloud/ipfs/QmcPsUASuDWuxbpoXPjgBrS6UHuxEdp3ZFwM88dKYsvTVH",
    "name": "mynamebrody #17877779234167542",
  },
  "name": "mynamebrody #17877779234167542",
  "networkId": 1,
  "tokenAddress": "0xF6793dA657495ffeFF9Ee6350824910Abc21356C",
  "tokenId": "30105810670306411381802272225292624615592980350863712708305530245480406056962",
  "tokenURI": "ipfs://ipfs/QmfHrsEpXXrvi2dTTNake723kkMapDQXeuYDsVmSRdsQNH",
  "tokenURL": "https://ipfs.fleek.co/ipfs/QmfHrsEpXXrvi2dTTNake723kkMapDQXeuYDsVmSRdsQNH",
  "tokenURLMimeType": "application/json",
}
`;
